File: 238_product_of_array_except_self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int pivot = 1;
        int len = nums.length;
        int result[] = new int[len];
        for (int i = 0; i < len; i++){
            result[i] = pivot;
            pivot *= nums[i];
        }

        pivot = 1;
        // backtrace from end of output
        for (int j = len - 1;j >= 0; j--){
            result[j] *= pivot;
            pivot *= nums[j];
        }
        return result;
    }
}

File: python/238_product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        N = len(nums)
        res = [1] * N
        for i in range(N - 1):
            res[i + 1] = res[i] * nums[i]

        pivot = 1

        for j in range(N - 1, -1, -1):
            res[j] *= pivot
            pivot *= nums[j]
        return res
